### PR TITLE
2839 time picker spinner improved keyboard input and added error state

### LINKED ISSUE #2839: [TimePickerSpinner] prevent invalid input and or add error state
### What package is this for?

- [x] React
- [ ] Angular

### Describe the bug
It is possible to add strings and invalid numbers to the TimePickerSpinner. If possible it should be prevented but if we can't prevent it we should add an error state and message like the NumberInput component has.

If we need an error state then we should make sure that the other components that depend on the TimePickerSpinner are updated to handle the error state.

### To Reproduce

1. Go to TimePickerSpinner
2. Type a string into the input

> Please create a reduced test case in CodeSandbox:
> https://codesandbox.io/s/github/IBM/carbon-addons-iot-react/tree/master/examples/codesandbox

### Expected behavior

No strings allowed, only valid numbers 

### Environment/versions:

- OS: [e.g. MacOS, Windows]
- Browser: [e.g. chrome, safari]
- `carbon-addons-iot-react` version: [e.g. `v2.60.0`]

### Additional context

image

### Specific timeline issues / requests

Do you want this work within a specific time period? Is it related to an
upcoming release?

_NB: The core contributors will try to work with your timeline, but it's not
guaranteed. The earlier you make a request in advance of a desired delivery
date, the better!_

### PR DESCRIPTION
**Pull request by bjornalm**
Closes #2839 
Improved keyboard input and added a validation state to the onChangeCallback for the TimePickerSpinner.

It turned our that completely preventing all illegal input without crippling the normal UX (text selection, cut, paste, tab focus etc) would be near impossible. So I made the TimePickerSpinner play well with the invalid prop and added the calculated validation result to the onChange callback so that it could be lifted to the parent component. Hence this is an optional built-in validation that still allow for external validation, or none, to be used.

**Change List (commits, features, bugs, etc)**
TimePickerSpinner
- added validation and adjusted scss
- kept a small part of the initial effort in preventing invalid input, i.e. to only allow numbers and one colon
- fixed old bx-prefix in scss
- fixed story knobs
- added new story
- added default i18n prop merge (turned out it was not needed for my changes but I left it in there)
- added unit tests

DateTimePicker/DateTimePicker2
- fix to disable apply button when there are invalid values in the TimePickerSpinner
- fixed issue in DateTimePicker where the timepickerspinner values from relative spilled over to absolute and the other way around.
- added new unit test and e2e tests
- added restore faketimer to tests 

As a note I also created the following new issues:
https://github.com/carbon-design-system/carbon-addons-iot-react/issues/2899
https://github.com/carbon-design-system/carbon-addons-iot-react/issues/2893
https://github.com/carbon-design-system/carbon-addons-iot-react/issues/2877

**Acceptance Test (how to verify the PR)**
- Verify that new [TimeSpinnerPicker](https://deploy-preview-XXXXX--carbon-addons-iot-react.netlify.app?path=/story/1-watson-iot-timepickerspinner--with-state-for-inputvalidation) story shows a validation error when an incorrect time is added. It should work for both 12 and 24 hours 

- Verify that it is not possible to, as an end user, enter an invalid time in the TimePickerSpinners we have in the DateTimePicker and DateTimePicker2 without getting the Apply button disabled. Also verify that the the UI with the validation error present looks ok.

**Regression Test (how to make sure this PR doesn't break old functionality)**
- Verify that the old TimeSpinnerPicker stories work as expected
- Verify that the DateTimePicker and DateTimePicker2 stories work as expected

**Things to look for during review**

- [ ] Make sure all references to `iot` or `bx` class prefix is using the prefix variable
- [ ] (React) All major areas have a `data-testid` attribute. New test ids should have test written to ensure they are not changed or removed.
- [ ] UI should be checked in RTL mode to ensure the proper handling of layout and text.
- [ ] All strings should be translatable.
- [ ] The code should pass a11y scans (The storybook a11y knob should show no violations). New components should have a11y test written.
- [ ] Unit test should be written and should have a coverage of 90% or higher in all areas.
- [ ] All components should be passing visual regression test. For new styles or components either a visual regression test should be written for all permutations or the base image updated.
- [ ] Changes or new components should either write new or update existing documentation.
- [ ] PR should link and close out an existing issue

### CODE REVIEW

File: packages/react/src/components/DateTimePicker/DateTimePickerV2.jsx [hunk 1]
```diff
@@ -283,6 +283,9 @@ const DateTimePicker = ({
   const [relativeValue, setRelativeValue] = useState(null);
   const [absoluteValue, setAbsoluteValue] = useState(null);
   const [focusOnFirstField, setFocusOnFirstField] = useState(true);
+  const [relativeToTimeInvalid, setRelativeToTimeInvalid] = useState(false);
+  const [absolutStartTimeInvalid, setAbsolutStartTimeInvalid] = useState(false);
+  const [absolutEndTimeInvalid, setAbsolutEndTimeInvalid] = useState(false);
```
Comment (kevinsperrine) [inline packages/react/src/components/DateTimePicker/DateTimePickerV2.jsx:288]: typo: `absolute`

File: packages/react/src/components/DateTimePicker/DateTimePicker.jsx [hunk 2]
```diff
@@ -309,6 +309,9 @@ const DateTimePicker = ({
   const [focusOnFirstField, setFocusOnFirstField] = useState(true);
   const [isTooltipOpen, setIsTooltipOpen] = useState(false);
   const [relativeLastNumberInvalid, setRelativeLastNumberInvalid] = useState(false);
+  const [relativeToTimeInvalid, setRelativeToTimeInvalid] = useState(false);
+  const [absolutStartTimeInvalid, setAbsolutStartTimeInvalid] = useState(false);
+  const [absolutEndTimeInvalid, setAbsolutEndTimeInvalid] = useState(false);
```
Comment (kevinsperrine) [inline packages/react/src/components/DateTimePicker/DateTimePicker.jsx:314]: typo: `absolute`

File: packages/react/src/components/DateTimePicker/DateTimePicker.jsx [hunk 3]
```diff
@@ -812,17 +816,31 @@ const DateTimePicker = ({
...
+  const onAbsoluteEndTimeChange = (pickerValue, evt, meta) => {
+    setAbsolutEndTimeInvalid(meta.invalid);
     changeAbsolutePropertyValue('endTime', pickerValue);
   };
 
   const tooltipValue = renderPresetTooltipText
     ? renderPresetTooltipText(currentValue)
     : getIntervalValue();
 
+  const disbableRelativeApply =
```
Comment (kevinsperrine) [inline packages/react/src/components/DateTimePicker/DateTimePicker.jsx:832]: typo: `disable`

File: packages/react/src/components/DateTimePicker/DateTimePicker.jsx [hunk 4]
```diff
@@ -812,17 +816,31 @@ const DateTimePicker = ({
...
+    isCustomRange &&
+    customRangeKind === PICKER_KINDS.RELATIVE &&
+    (relativeLastNumberInvalid || relativeToTimeInvalid);
+
+  const disbableAbsoluteApply =
```
Comment (kevinsperrine) [inline packages/react/src/components/DateTimePicker/DateTimePicker.jsx:837]: typo: `disable`

File: packages/react/src/components/DateTimePicker/DateTimePickerV2.jsx [hunk 5]
```diff
@@ -629,17 +633,29 @@ const DateTimePicker = ({
...
+  const onAbsoluteEndTimeChange = (pickerValue, evt, meta) => {
+    setAbsolutEndTimeInvalid(meta.invalid);
     changeAbsolutePropertyValue('endTime', pickerValue);
   };
 
   const tooltipValue = renderPresetTooltipText
     ? renderPresetTooltipText(currentValue)
     : getIntervalValue();
 
+  const disbableRelativeApply =
```
Comment (kevinsperrine) [inline packages/react/src/components/DateTimePicker/DateTimePickerV2.jsx:649]: typo: `disable`

File: packages/react/src/components/DateTimePicker/DateTimePickerV2.jsx [hunk 6]
```diff
@@ -629,17 +633,29 @@ const DateTimePicker = ({
...
+    isCustomRange && customRangeKind === PICKER_KINDS.RELATIVE && relativeToTimeInvalid;
+
+  const disbableAbsoluteApply =
```
Comment (kevinsperrine) [inline packages/react/src/components/DateTimePicker/DateTimePickerV2.jsx:652]: typo: `disable`

**Review by kevinsperrine — APPROVED** [review]
Nice work! That's a good improvement. Just a few typo comments, and one edge case. If you enter an invalid time, and then turn off time input, it doesn't clear the invalid state, and you can't click apply. Not sure if that's a case worth fixing or not, but everything else looks great.

File: packages/react/src/components/DateTimePicker/DateTimePicker.jsx (hunk 4 above)
Comment (davidicus) [reply packages/react/src/components/DateTimePicker/DateTimePicker.jsx:837]: +1

File: packages/react/src/components/DateTimePicker/DateTimePicker.jsx [hunk 7]
```diff
@@ -812,17 +816,31 @@ const DateTimePicker = ({
...
   const tooltipValue = renderPresetTooltipText
     ? renderPresetTooltipText(currentValue)
     : getIntervalValue();
 
+  const disableRelativeApply =
+    isCustomRange &&
+    customRangeKind === PICKER_KINDS.RELATIVE &&
+    (relativeLastNumberInvalid || relativeToTimeInvalid);
+
+  const disbableAbsoluteApply =
```
Comment (davidicus) [inline packages/react/src/components/DateTimePicker/DateTimePicker.jsx:837]:
```suggestion
  const disableAbsoluteApply =
```

File: packages/react/src/components/DateTimePicker/DateTimePicker.jsx [hunk 8]
```diff
@@ -812,17 +816,31 @@ const DateTimePicker = ({
...
+    isCustomRange &&
+    customRangeKind === PICKER_KINDS.ABSOLUTE &&
+    (absoluteStartTimeInvalid || absoluteEndTimeInvalid);
+
+  const disableApply = disableRelativeApply || disbableAbsoluteApply;
```
Comment (davidicus) [inline packages/react/src/components/DateTimePicker/DateTimePicker.jsx:842]:
```suggestion
  const disableApply = disableRelativeApply || disableAbsoluteApply;
```

File: packages/react/src/components/DateTimePicker/DateTimePickerV2.jsx [hunk 9]
```diff
@@ -629,17 +633,29 @@ const DateTimePicker = ({
...
   };
 
   const tooltipValue = renderPresetTooltipText
     ? renderPresetTooltipText(currentValue)
     : getIntervalValue();
 
+  const disableRelativeApply =
+    isCustomRange && customRangeKind === PICKER_KINDS.RELATIVE && relativeToTimeInvalid;
+
+  const disbableAbsoluteApply =
```
Comment (davidicus) [inline packages/react/src/components/DateTimePicker/DateTimePickerV2.jsx:652]:
```suggestion
  const disableAbsoluteApply =
```

File: packages/react/src/components/DateTimePicker/DateTimePickerV2.jsx [hunk 10]
```diff
@@ -629,17 +633,29 @@ const DateTimePicker = ({
...
+    isCustomRange &&
+    customRangeKind === PICKER_KINDS.ABSOLUTE &&
+    (absoluteStartTimeInvalid || absoluteEndTimeInvalid);
+
+  const disableApply = disableRelativeApply || disbableAbsoluteApply;
```
Comment (davidicus) [inline packages/react/src/components/DateTimePicker/DateTimePickerV2.jsx:657]:
```suggestion
  const disableApply = disableRelativeApply || disableAbsoluteApply;
```

**Review by davidicus — APPROVED** [review]

**Pull request merged**